### ISSUE TITLE
Bug: [no-unnecessary-condition] does not work properly with void

### ISSUE THREAD

**Issue by Zamiell** [typescript-eslint/typescript-eslint#5254]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3Abug+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVJxwAoBKALnlWQFsAjEGeAH3gDcctgBuAWACh+YPAGcMBHACUQw5BDEBecaV58hqUfDgZkMVCGABBAHI16jRYkJSZc+PPvw0oRFj3d+-IA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1tiacTJTIAhtEK0yHJgBNK+SpPRQA7iKaRwYAL4htQA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA

**Comment by bradzacher** [comment]
This is a bug - though as per #5255 - you really should avoid using `| void`

**Comment by Zamiell** [comment]
If the `consistent-return` ESLint rule is turned on, I'm not quite sure I understand the rational for `no-invalid-void-type`.

For example:

```ts
type FooReturnType: number | void;

function foo1(): FooReturnType{
  // Do some stuff.
  // In this function, I'm not returning anything, so I can avoid typing a return statement at all to keep things nice and clean.
}

function foo2(): FooReturnType {
  if (someCondition) {
    return 123;
  }

  // Base case - return nothing.
  // - If I don't include a return statement, then "consistent-return" complains.
  // - If I just include a "return" statement, then "consistent-return" complains.
  // - Thus, I must explicitly use a "return undefined" statement.
  return undefined;
}
```

**Comment by bradzacher** [comment]
With TS if you turn on [`noImplicitReturns`](https://www.typescriptlang.org/tsconfig#noImplicitReturns) then you don't need `consistent-return` as TS will enforce the same thing, but do it in a type-aware way.

If you start explicitly unioning `void` in your return types instead of `undefined`, however, TS won't be able to help you because it accepts that `void` needs no return. [repl](https://typescript-eslint.io/play/#ts=4.7.2&sourceType=module&code=C4TwDgpgBAQghgEwGIHsUCULAK4CcB2AKuNALxT7YC2ARhLlAD5QBuKAlggNwBQPAZtnwBjYOxT4o-NAEYAFAEoAXLESoMWPERJQA3jyhQA9EYooouTQQoQICO1AgsIk4AAsU2AOZuo7+tDsAM5QcJIQAB6QonBiEn5uwVJCouKSALIAkgDiABKEFlaScBTUdLg8AL58giJxktIoAExyQShUEADCEgjs9So0aAA2EGHKqshomDgExJB6BlDs-FCt7V09fWkKC4aGljOSMk0AzLyG1YsmUIQAylAIKBAh+CjANg4lB1qOzq4e3l8AHdoAAqIFhYCgqDCCRBYLAFzvb4EIJXUyw-DwoKI-DAAC0KPCuFwKFwaMuoHm2TQkw0hzmZFKtHoTCgQns-HY+DsvAEKXqUjQJ0UKhpKDp0y0jN2xlMd0cUQgohCXyKUDowjg2CC0BBUDccGcoTM+HxbE4hUOfhI6JhcIRSMJ6sezzM7x5nytWiqNQFaSFKAALGsOt18L1+hrhqN8ONxZKijL9IZlqs2mHNvUdim9kSoMczotLoZrgrItFgKrvdZNdrdVB9YbjSVXmaLQga65baWMQ6cU783Ahm1FZXq0TfTwy4kQkkgW4AvasY68c7rUFsMIANYhIF9XwK2H2JQ8FzUG7Pd66KAAQQANLBH50oNVaql4o0AKxyOC4LwqIQV45osQT7sAwi+L+-4gXsMJwA2QE4gAdLep5wfs6oyHycFaohV7ITA6EYfmTQ4XseHQEhwDIZ0xFwfmRYXHwvY3Pc26vECIT+FAYF9JBSwhJEho6mILYRgkOqhEMQwargYSQW6KArDx76Cn+0Cws4lgIHamLYriBL5q6LxvOyEb0DiYSdjxfEQb4STCfWYnQNZkkhNIuDCG6ICeHacHAOYiCdmE5k8t5QRBH+ICPkIlhwJBcA0CMXZVEAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQMYHsA7ZAS2QBdFDyBaaRc2aQ9KRaafaScMAXxB8gA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3QAacDUhFYASSSomyPACV0eNplIUwo9FIC+IQ0A)

This is why you should use the rule - you should not use `void` unless it's by itself in a return type location.

**Comment by Zamiell** [comment]
Thanks Brad, that makes a lot of sense.

Regarding your final example, I use Airbnb rules, which mandates the `default-case` ESLint rule. So I typically use a `ensureAllCases` there:

```ts
export const ensureAllCases = (obj: never): never => obj;
```

Doing this obviates the needs for TS to make the return type exhaustive, since the switch becomes exhaustive. Which I guess would be considered worse, but since I have to make a default case anyway, it seems like a good idea. I imagine that you might counter that I should turn off the `default-case` rule entirely since it sucks in TypeScript. (And I'd be open to that, although I have historically found the rule quite useful in situations where something isn't being returned from the switch statement.)

Furthermore, in my notes, I have written down that `consistent-return` catches more bugs and is more exhaustive than `noImplicitReturns`. But I don't recall specifically why I wrote this. Do you know any cases where this is true?

For now, I'll try turning the rule off and see how things go. :)

**Comment by bradzacher** [comment]
> I imagine that you might counter that I should turn off the default-case rule entirely since it sucks in TypeScript.

Yeah, exactly correct - `default-case` is another example of a *generally* "bad" rule for TS code because again it's not type-aware; so it doesn't understand that a switch can be exhaustive without the need for a `default`. Our [`switch-exhaustiveness-check`](https://typescript-eslint.io/rules/switch-exhaustiveness-check) is a better type-aware alternative in that regard.

There are *some* cases where it can help enforce good code; like if you're `switch`ing on a non-union type:
```ts
declare const someString: string;
switch (someString) {
  case 'someValue1': { ... }
  case 'someValue2': { ... }
  // ...
  // really should have a default to handle the 'other' case
}
```
But from experience `switch`ing on a pure `string`/`number` is pretty rare to to in good TS code because if you're writing out cases then you already have the basis for a union type!

----

> Furthermore, in my notes, I have written down that consistent-return catches more bugs and is more exhaustive than noImplicitReturns. But I don't recall specifically why I wrote this. Do you know of any cases where this is true?

I don't - I'm sure there would be some because TS is driven by types but the lint rule is driven purely by syntax. Personally I've found TS to cover all the cases I want for codebase standardisation!

**Comment by Zamiell** [comment]
Brad,

Thanks for the reply. I've turned off `consistent-return`, but I've noticed that the following erroneous pattern is now undetected:

```ts
function foo() { // Implicit return type
  if (someCondition) {
    return;
  }

  return undefined;
}
```

And, very similarly:

```ts
function foo(): undefined { // Explicit return type
  if (someCondition) {
    return;
  }

  return undefined;
}
```

Here, either implicitly or explicitly, the return lines are heterogeneous. Previously, the `consistent-return` rule would force me to change the `return;` line to `return undefined;`, which makes sense - all the paths outside of the function should be the same.

In this situation, I am stuck, because I can't turn `consistent-return` back on without ruining the `foo5` case that you outlined in your previous reply. Is there some separate lint rule other than `consistent-return` that I should use to handle this?
